### PR TITLE
Active Learning Methods

### PR DESCRIPTION
**Pull request by nitish1295**
## Description

The code base contains the following figure:

1.  2D Classification dataset
2.  1D Regression Dataset 
3. MNIST Plot to indicates probabilities and uncertainties of samples queried by active learning
4. Comparison of QBC, uncertainty sampling and random sampling approaches for 
       a) classification problem
       b) regression problem

6. Decision Countor for displaying queried samples via a uncertainty sampling vs a random sampling approach for classification on same initial train data
7. Regression Lines for displaying queried samples via a uncertainty sampling vs a random sampling approach for classification on same initial train data

### Figure Number

In Progress

### Figures

| 1.  and 4(a)|  image| image|
|:-: | :-: | :-: |
|2.  and 4(b) |image| image|

---

|6|image |
| :-: | :-: |
| 7.|image|
|3.| image|

### Issue 

- https://github.com/probml/pyprobml/issues/1016

## Checklist:

- [x] Performed a self-review of the code
- [x] Tested on Google Colab.

## Potential problems/Important remarks

- As per inputs from @patel-zeel will remove the last row for Fig 3. Will try adjust the Fig 3. MNIST plot to fit in the given fig_height=1.5, earlier this was not the case since it was not possible to fit all the 6 plots in fig_height=1.5 without making them unreadable. I understand that there are commented debug statements where the MNIST plot is created, I will remove them as well
- Currently the Decision Countor plots are saved for all the different query strategies, will try to ensure that they are saved only for the required query stratergies
- In order to accomodate celluloid, will need ffmpeg for the github workflow else the tests will fail. Link to github actions [FFmpeg](https://github.com/marketplace/actions/setup-ffmpeg)

### CODE REVIEW

**Comment by patel-zeel** [comment]
@nitish1295 Can you break this notebook into multiple notebooks so that it is easier for Dr. Murphy to review?

**Comment by nitish1295** [comment]
@patel-zeel Oh okay, I actually use the same functions over and over for different plots hence in order to prevent repetition of code I kept it in one notebook, let me know if there is a way around it? Or let me know how you want me to break this, maybe accuracy and RMSE plots in one notebook and the rest of the code in another notebook?

**Comment by nipunbatra** [comment]
Should we perhaps keep the animation code outside the main notebook in pyprobml to ensure that we do not need to install additional dependencies? Or is the overhead of installing ffmpeg minimal? 

The animation code could go in a notebook that need not be run by the Github Actions.

**Comment by patel-zeel** [comment]
We can add repetitive functions to probml-utils.

**Comment by nitish1295** [comment]
The github actions would require https://github.com/marketplace/actions/setup-ffmpeg, as for my code right now the animations and the plotting functions are not completely decoupled. 

Usage for ffmpeg
image

Another way could be that we could have an environment variable in github workflow which can indicate that the code is running under a github workflow and based on that I can skip animation display

**Comment by nitish1295** [comment]
Few things to note:

- I have updated the MNIST figure without the last row(uncertainty) in the overleaf template, please note that now 0 is chosen instead of 3 by margin sampling my best guess is this happened because I did not have a random state set for logistic regression. The spacing issues in second row are due to fig_width=6, I have tried using fig_width=4.2 but then the annotated text cuts the y-axis of the plot let me know if you have any inputs, I have also tried the following ```plt.subplots_adjust(wspace=0.2)``` but didn't help much either. I will try a few more things and let you know
- Another thing I noticed is that if I call the pml.latexify() function multiple times it does not update the parameters for the initial call. As an example my code uses MNIST with ```pml.latexify(fig_width=6, fig_height=1.5)``` and after that for my regression plots I call ```pml.latexify(width_scale_factor=3)``` multiple times but my regression plots are saved as follows:

| fig counter| fig_width  | fig_height |
|:-: | :-: | :-: |
|1| 6 | 1.5|
|2| 2 | 1.5|
|3| 2 | 1.5|
|4| 2 | 1.5|

The first plot follows the same old config. This happens irrespective of if I run the code manually or run it using ```pytest -n auto -s --verbose tests/test_notebooks.py```. Also if I execute the regression plot cell again manually everything works fine

**Comment by patel-zeel** [comment]
I think the latexify problem will be solved once you split your notebook into multiple parts.

for the FFmpeg issue, we can either ask the workflow not to run the animation code using an environment variable named "WORKFLOW_EXECUTION" or put the animation code in a separate notebook under the `notebooks/book2/34/supplementary` folder. You can choose and let me know whichever way is easier for you.

**Comment by nitish1295** [comment]
As discussed there is a default environment variable which can help us with our requirement already in the Github workflow. It is available at every step in the workflow



GITHUB_ACTIONS | Always set to true when GitHub Actions is running the workflow. You can use this variable to differentiate when tests are being run locally or by GitHub Actions.
-- | --



</body>
</html>

You can read more about it at https://docs.github.com/en/actions/learn-github-actions/environment-variables

**Comment by nitish1295** [comment]
Few points:

### MNIST Plot Changes

Since I will change the MNIST plot for the second time today, here is the explanation

The logistic regression code in scikit-learn is placed at [scikit-learn](https://github.com/scikit-learn/scikit-learn/tree/baf0ea25d6dd034403370fea552b21a6776bef18)/[sklearn](https://github.com/scikit-learn/scikit-learn/tree/baf0ea25d6dd034403370fea552b21a6776bef18/sklearn)/[linear_model](https://github.com/scikit-learn/scikit-learn/tree/baf0ea25d6dd034403370fea552b21a6776bef18/sklearn/linear_model)/_logistic.py /. 

* Google Colab uses version 1.0.2
* On local machine I was using 1.1.1 and Python 3.8(Installed it earlier on my machine)

And they give different results for MNIST Plot under margin sampling. In Colab margin sampling chooses the point 3 as most uncertain while on my local system we get point 0 as the most uncertain. **Once I reverted back to scikit-learn==1.0.2, the MNIST plot was the same**.  

Furthermore there are significant differences between the above mentioned file which contains the code for logistic regression, comparison between the 2 versions --> https://github.com/scikit-learn/scikit-learn/compare/1.0.2...1.1.1. You can search for **sklearn/linear_model/_logistic.py** using Ctrl+F on the page and you should come across the file mentioned above which has significant changes, although I was unable to point the exact one

### ffmpeg dependency
The ffmpeg dependency was not a bi product of using celluloid but rather the way I was displaying the animation, this [Stackoverflow question](https://stackoverflow.com/questions/25333732/matplotlib-animation-not-working-in-ipython-notebook-blank-plot) has a very detailed explanation for using animation with matplotlib. From the following 2 options I have changed to the second one:

1. HTML(animation.**to_html5_video()**) - Requires ffmpeg
2. HTML(animation.**to_jshtml()**) - This does not require ffmpeg, ran it on local machine (FYI I didn't know about this else I would've used it from start)

Now animations will look as follows:
image

I think they are slightly better than before and I can live with this, let me know what you think
